### PR TITLE
Create Migration Guide for Cordova 5.0.0-beta1

### PR DESCRIPTION
**Pull request by jennantilla**
# Description
## One Line Summary
Create Migration Guide for Cordova 5.0.0-beta1

### Motivation
Facilitate upgrade to User Model.

### OPTIONAL - Other
Models after Native [iOS](https://github.com/OneSignal/OneSignal-iOS-SDK/blob/major_release_5.0.0/MIGRATION_GUIDE.md) and [Android](https://github.com/OneSignal/OneSignal-Android-SDK/blob/5.0.0-beta1/MIGRATION_GUIDE.md) migration guides with relevant info for both platforms/SDKs.

# Testing
Updated code examples tested in Cordova app.

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [X] Public API changes

# Checklist
## Overview
   - [X] I have filled out all **REQUIRED** sections above
   - [X] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [X] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [X] I have included test coverage for these changes, or explained why they are not needed
   - [X] All automated tests pass, or I explained why that is not possible
   - [X] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [X] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [X] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

- - -
This change is [Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Cordova-SDK/864)

### CODE REVIEW

**Comment by jennantilla** [comment]
@iAmWillShepherd adding you as a reviewer to the Cordova 5.0.0-beta1 migration guide. Feel free to provide any feedback!

I also spun up a test app that I used when developing the 5.0.0 plugin. Do you think it might be a helpful resource for dev-rel?

https://github.com/jennantilla/onesignal-cordova-example

File: MIGRATION_GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+| `let canRequest = window.plugins.OneSignal.Notifications.canRequestPermission();`  | *Whether attempting to request notification permission will show a prompt. Returns `true` if the device has not been prompted for push notification permission already.*                                                                                                                                                                                                                                                |
+| `window.plugins.OneSignal.Notifications.clearAll();`                                           | *Removes all OneSignal notifications.*|                                                                                                                                                           
+| `window.plugins.OneSignal.Notifications.removeNotification();`                                                                                               | *(Android only) Cancels a single OneSignal notification based on its Android notification integer ID. Use instead of Android's [android.app.NotificationManager.cancel], otherwise the notification will be restored when your app is restarted.*|
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:256]: removeNotification needs an argument, the notification ID.

File: MIGRATION_GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+| `window.plugins.OneSignal.Notifications.removeNotification();`                                                                                               | *(Android only) Cancels a single OneSignal notification based on its Android notification integer ID. Use instead of Android's [android.app.NotificationManager.cancel], otherwise the notification will be restored when your app is restarted.*|                                                                                                                                                                                                    
+| `window.plugins.OneSignal.Notifications.removeGroupedNotifications();`                                                                                               | *(Android only) Cancels a group of OneSignal notifications with the provided group key. Grouping notifications is a OneSignal concept, there is no [android.app.NotificationManager] equivalent.*|
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:257]: removeGroupedNotifications needs an argument

File: MIGRATION_GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+| `window.plugins.OneSignal.Notifications.setNotificationWillShowInForegroundHandler();`<br><br>***See below for usage***       | *Sets the handler to run before displaying a notification while the app is in focus. Use this handler to read notification data and change it or decide if the notification ***should*** show or not.<br><br>***Note:*** this runs ***after*** the [Notification Service Extension](https://documentation.onesignal.com/docs/service-extensions) which can be used to modify the notification before showing it.* |
+| `(void)setNotificationOpenedHandler:(OSNotificationOpenedBlock _Nullable)block`<br><br>***See below for usage***                   | *Sets a handler that will run whenever a notification is opened by the user.*|                                                                                                                                                                                                                                                                                                                     
+
+### Prompt for Push Notification Permission
+
+**Cordova/Ionic**
+```typescript
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:268]: nit: indentation?

File: MIGRATION_GUIDE.md [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+| `window.plugins.OneSignal.Notifications.registerForProvisionalAuthorization();`                  | *(iOS only) Instead of having to prompt the user for permission to send them push notifications, your app can request provisional authorization.*|                                                                                                                                                                                                    
+| `window.plugins.OneSignal.Notifications.addPermissionObserver();`<br><br>***See below for usage***                   | *The `OSPermissionObserver.onOSPermissionChanged` method will be fired on the passed-in object when a notification permission setting changes. This happens when the user enables or disables notifications for your app from the system settings outside of your app.*|
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:260]: `addPermissionObserver` needs a parameter?

> The OSPermissionObserver.onOSPermissionChanged method will be fired on the passed-in object when a notification permission setting changes.

^ Does not apply to cordova sdk

File: MIGRATION_GUIDE.md [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+| `window.plugins.OneSignal.Notifications.removePermissionObserver(observer1);`<br><br>***See below for usage***        | *Remove a push permission observer that has been previously added.*|                                                                                                                                                                                                                         
+| `window.plugins.OneSignal.Notifications.setNotificationWillShowInForegroundHandler();`<br><br>***See below for usage***       | *Sets the handler to run before displaying a notification while the app is in focus. Use this handler to read notification data and change it or decide if the notification ***should*** show or not.<br><br>***Note:*** this runs ***after*** the [Notification Service Extension](https://documentation.onesignal.com/docs/service-extensions) which can be used to modify the notification before showing it.* |
+| `(void)setNotificationOpenedHandler:(OSNotificationOpenedBlock _Nullable)block`<br><br>***See below for usage***                   | *Sets a handler that will run whenever a notification is opened by the user.*|
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:263]: This is objc syntax?

File: MIGRATION_GUIDE.md [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+**Cordova/Ionic**
+```typescript
+    window.plugins.OneSignal.Notifications.requestPermission(function(accepted) {
+    console.log("User accepted notifications: " + accepted);
+});
+```
+
+### Permission Observer
+
+Any object implementing the `OSPermissionObserver` protocol can be added as an observer. You can call `removePermissionObserver` to remove any existing listeners.
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:276]: > Any object implementing the OSPermissionObserver protocol can be added as an observer

This is how it works on native iOS but on cordova SDK, its just a function

File: MIGRATION_GUIDE.md [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+| `window.plugins.OneSignal.Notifications.removePermissionObserver(observer1);`<br><br>***See below for usage***        | *Remove a push permission observer that has been previously added.*|                                                                                                                                                                                                                         
+| `window.plugins.OneSignal.Notifications.setNotificationWillShowInForegroundHandler();`<br><br>***See below for usage***       | *Sets the handler to run before displaying a notification while the app is in focus. Use this handler to read notification data and change it or decide if the notification ***should*** show or not.<br><br>***Note:*** this runs ***after*** the [Notification Service Extension](https://documentation.onesignal.com/docs/service-extensions) which can be used to modify the notification before showing it.* |
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:262]: setNotificationWillShowInForegroundHandler needs a parameter

File: MIGRATION_GUIDE.md [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+```
+
+## Location Namespace
+
+The Location namespace is accessible via `OneSignal.Location` and provide access to location-scoped functionality.
+
+| **Cordova/Ionic**                                         | **Description**                                                                           |
+| ----------------------------------------------------------|-------------------------------------------------------------------------------------------|
+| `window.plugins.OneSignal.Location.isShared();` <br><br>***See below for usage***             | *Whether location is currently shared with OneSignal.*|
+| `window.plugins.OneSignal.Notifications.requestPermission();` | *Use this method to manually prompt the user for location permissions. This allows for geotagging so you send notifications to users based on location.* |
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:314]: oops not `Notifications` namespace

File: MIGRATION_GUIDE.md [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+    
+    // Add SMS subscription
+    window.plugins.OneSignal.User.addSms("+15558675309");
+    // Remove previously added SMS subscription
+    window.plugins.OneSignal.User.removeSms("+15558675309");
+```
+
+# API Reference
+
+Below is a comprehensive reference to the `5.0.0-beta1` OneSignal Cordvoa SDK.
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:140]: nit spelling

File: MIGRATION_GUIDE.md [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+| `window.plugins.OneSignal.InAppMessages.clearTriggers();`                                                                                  | *Clear all triggers from the current user.*                                                                                                                                                                                                                                                                                                                                                                                                                                     |
+| ` window.plugins.OneSignal.InAppMessages.setLifecycleHandler();`<br><br>***See below for usage*** | *Set the in-app message lifecycle handler.*                                                                                                                                                                                                                                                                                                                                                                                                                                 |
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:336]: setLifecycleHandler and setClickHandler needs a parameter
Is there a `setLifecycleHandler` example code snippet?

File: MIGRATION_GUIDE.md [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+| `window.plugins.OneSignal.User.removeEmail("customer@company.com");`                             | *Remove an email subscription from the current user. Returns `false` if the specified email does not exist on the user within the SDK, and no request will be made.*                                                               |
+| `window.plugins.OneSignal.User.addSms("+15558675309");`                                                   | *Add a new SMS subscription to the current user.*                                                                                                                                                                                        |
+| `window.plugins.OneSignal.User.removeSmsNumber("+15558675309");`                                 | *Remove an SMS subscription from the current user. Returns `false` if the specified SMS number does not exist on the user within the SDK, and no request will be made.*                                                            |
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:195]: I think its now just `removeSms`?

File: MIGRATION_GUIDE.md [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+
+
+
+## Notifications Namespace
+
+The Notifications namespace is accessible via `OneSignal.Notifications` and provides access to notification-scoped functionality.
+
+| **Cordova/Ionic**                                                         |                                                                                              **Description** |
+|---------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------|
+| `let persmission = window.plugins.Notifications.permission`               | *Whether this app has push notification permission.*                                                                                                                                                                                                                                                                                                                                                                        |
```
Comment (nan-li) [inline MIGRATION_GUIDE.md:253]: nit spelling persmission

**Review by nan-li — APPROVED** [review]
LGTM

**Pull request merged**